cmd/scan_user.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"go-module/dbtools"
+	"go-module/explorerScan"
+	"go-module/model"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+var (
+	wg  sync.WaitGroup
+	mux sync.Mutex
+
+	supportedDApps = map[string]string{ // project_code: pool_type
+		"bsc_apeswap":     "liquidity",
+		"bsc_coinswap":    "common",
+		"bsc_biswap":      "common",
+		"bsc_babyswap":    "liquidity",
+		"bsc_mdex":        "liquid",
+		"bsc_pancakeswap": "farming",
+	}
+
+	maxRoutineGroup = 20
+)
+
+func main() {
+	indexLogsLiquidityData()
+}
+
+func indexLogsLiquidityData() {
+	//Get pool list from DB
+	err := dbtools.DBInitializer()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Init explorerScan API service
+	api := explorerScan.Init()
+	go api.HandleStateMachine()
+
+	nextRoutine := make(chan int, maxRoutineGroup)
+	for i := 0; i < maxRoutineGroup; i++ {
+		nextRoutine <- i
+	}
+
+	for projectCode, poolType := range supportedDApps {
+
+		startTime := time.Now().Unix()
+		pools := dbtools.GetPoolsList(projectCode, poolType)
+		poolNum := len(pools)
+		fmt.Printf("Total %d pools\n", poolNum)
+		poolRemaining := poolNum
+		lastBlock := getLastBlock()
+
+		// Create Routine to handle each pool
+		for i := 0; i < poolNum; i++ {
+			pool := pools[poolNum-i-1]
+			components := dbtools.GetComponentsOfPool(pool.Address)
+
+			<-nextRoutine
+			wg.Add(1)
+			go func(id int, Pool model.Pool, Components []model.PoolComponent) {
+				defer wg.Done()
+
+				holdersLiquidity := api.GetLiquidityData(pool, lastBlock, components)
+				nextRoutine <- 0
+
+				mux.Lock()
+
+				if len(holdersLiquidity) > 0 {
+					dbtools.CreateTable(&model.Holder{})
+					dbtools.Save(holdersLiquidity)
+				}
+
+				// Logger status
+				poolRemaining--
+				fmt.Println("")
+				fmt.Printf("Pool %s, %d results\n", pool.Address, len(holdersLiquidity))
+				fmt.Printf("%s, %d pools remaining\n", projectCode, poolRemaining)
+				fmt.Println("")
+
+				mux.Unlock()
+
+			}(i, pool, components)
+		}
+		wg.Wait()
+
+		// Log scan history
+		finishTime := time.Now().Unix()
+		dbtools.CreateTable(&model.ScanHistory{})
```

Comment:
```suggestion
		dbtools.CreateTable(&model.ScanHistory{})
```

Không chạy trong runtime code của chương trình mà cần chạy seeding riêng

Reply:
Comment này em chưa hiểu lắm. Tức là mình cần tạo file .json rồi chạy seeding giống phần pool với components đúng không ạ ??

Reply:
Em cần tạo model ở bên depocket databases và tạo script migration riêng.
Không tạo table hay migrate db ở trong code bên này